pkg/docker/build.go

```diff
@@ -21,9 +24,39 @@ RUN curl -o docker.tgz https://download.docker.com/linux/static/stable/x86_64/do
 COPY . $BUNDLE_DIR
 `
 
+// BuildInput represents stdin passed to the mixin for the build command.
+type BuildInput struct {
+	Config MixinConfig
+}
+
+// MixinConfig represents configuration that can be set on the docker mixin in porter.yaml
+// mixins:
+// - docker:
+//	  version: 19.03.8
+type MixinConfig struct {
+	ClientVersion string `yaml:"clientVersion,omitempty"`
+}
+
 // Build will generate the necessary Dockerfile lines
 // for an invocation image using this mixin
 func (m *Mixin) Build() error {
-	fmt.Fprintf(m.Out, dockerfileLines)
+	var input BuildInput
+	err := builder.LoadAction(m.Context, "", func(contents []byte) (interface{}, error) {
+		err := yaml.Unmarshal(contents, &input)
+		return &input, err
+	})
+	if err != nil {
+		return err
+	}
+
+	if input.Config.ClientVersion != "" {
+		m.DockerVersion = input.Config.ClientVersion
+	} else {
```

Comment:
Ah, I see the issue here.  Although this `else` is certainly fine, we can avoid it.

The reason we weren't seeing the default version unit test pass was b/c the `NewTestMixin()` helper didn't actually call `mixin.New()` which performs the action of setting the default.  In the process of adjusting that, I also updated the `mixin.New()` method signature to remove returning an `error`, to simplify logic.

Check it out: https://github.com/gaurimadhok/porter-docker/compare/set-version...vdice:gaurimadhok-set-version-plus-default-fix?expand=1

Feel free to use some/all of those changes.

Reply:
thank you that makes sense!